### PR TITLE
Add etcd-operator seamless integration with TLS

### PR DESCRIPTION
**Pull request by aanm**
EDIT: It turns out there's no need to perform code-changes. I'll used our current etcd-configuration map and mount the `cilium-etcd-client-tls` secrete into the Cilium pod.

**Summary of changes**:

Add seamless integration for etcd-operator running with TLS.

A developer cilium.yaml is available in `examples/kubernetes/addons/etcd-operator/cilium-development.yaml` which contains the necessary options to run etcd-operator.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4921)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: cilium/cmd/kvstore.go
Comment (aanm) [inline cilium/cmd/kvstore.go]: This is a leftover, it needs to be fixed as when using `cilium kvstore get` the cilium CLI command does not have the necessary configuration to contact kubernetes, to retrieve the certificates to contact etcd.

File: pkg/kvstore/etcd_tls.go
Comment (tgraf) [inline pkg/kvstore/etcd_tls.go]: Let's define constants for this as this is a stable API. No need to add dependencies.

File: examples/kubernetes/addons/etcd-operator/cilium-development.yaml
Comment (tgraf) [inline examples/kubernetes/addons/etcd-operator/cilium-development.yaml]: Duplicate kube-system

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/k8s/client/logfields.go
Comment (ianvernon) [inline pkg/k8s/client/logfields.go]: Should `LogSubsys` be changed due to this refactoring?

File: cilium/cmd/kvstore.go
Comment (ianvernon) [reply cilium/cmd/kvstore.go]: A few questions: 
* Is the kubeconfig always guaranteed to be at this path? I presume yes because we are mounting it as such in the Cilium DS, but this is user-configurable, yes?
* This will always result in `k8sclient.IsEnabled()` to return true:

https://github.com/cilium/cilium/blob/db674150168541c2d1dc6524d761f7e2d0f00906/pkg/k8s/config.go#L60-L66

`IsEnabled` returns true if the KubeconfigPath is non-nil, and you unconditionally set KubeconfigPath above in your call to `k8sclient.Configure`:

https://github.com/cilium/cilium/blob/db674150168541c2d1dc6524d761f7e2d0f00906/pkg/k8s/config.go#L48-L58

Won't this break non-K8s setups?

**Review by ianvernon — CHANGES_REQUESTED** [review]
Still reviewing, have one important observation that I'd like clarification on.

File: pkg/kvstore/etcd.go
Comment (ianvernon) [inline pkg/kvstore/etcd.go]: Is this guaranteed to always be `kube-system`?

File: pkg/kvstore/etcd.go
Comment (ianvernon) [inline pkg/kvstore/etcd.go]: Add a unit test for this function.

File: examples/kubernetes/addons/etcd-operator/cilium-development.yaml
Comment (ianvernon) [inline examples/kubernetes/addons/etcd-operator/cilium-development.yaml]: This can be removed since https://github.com/cilium/cilium/issues/1202 has been closed

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (ianvernon) [inline examples/kubernetes/addons/etcd-operator/README.md]: Use JSON parsing instead of grep / cut; this has proven brittle for us in the past due to the K8s CLI formatting changes.

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (ianvernon) [inline examples/kubernetes/addons/etcd-operator/README.md]: Should we change this to zero restarts? ;)

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: cilium/cmd/kvstore.go
Comment (raybejjani) [reply cilium/cmd/kvstore.go]: I think the path is returned by `/config` in https://github.com/cilium/cilium/blob/master/api/v1/models/daemon_configuration_status.go#L33:2 (I noticed this being printed in https://github.com/cilium/cilium/blob/master/cilium/cmd/config.go#L80:59)

That might simplify things a bit?

**Comment by aanm** [comment]
test-me-please

File: examples/kubernetes/addons/etcd-operator/tls/certs/ca-csr.json
Comment (ianvernon) [inline examples/kubernetes/addons/etcd-operator/tls/certs/ca-csr.json:1]: How do the contents of this commit match the commit title ?

**Comment by aanm** [comment]
test-me-please

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/README.md]: @ianvernon this is actually expected because kube-dns will keep restarting until the user changes the label

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/README.md]: Will do

File: examples/kubernetes/addons/etcd-operator/tls/certs/ca-csr.json
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/tls/certs/ca-csr.json:1]: Fixed

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-k8s-upstream

**Comment by aanm** [comment]
test-upstream-k8s

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (tgraf) [inline examples/kubernetes/addons/etcd-operator/README.md:59]: Let's deploy the operator first so we avoid the error below.

File: daemon/k8s_watcher.go
Comment (tgraf) [inline daemon/k8s_watcher.go]: Cluster member is missing

File: pkg/node/manager.go
Comment (tgraf) [inline pkg/node/manager.go]: This should check if the source of the old node is not Kubernetes. We still want to support updates of k8s node if no kvstore entry has been received so far

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by eloycoto** [comment]
@aanm I think that the jsonpath that you ask can be in different ways:

To retrieve only one: 
```
 kubectl -n kube-system get pods -l k8s-app=kube-dns -o jsonpath='{.items[0].metadata.name}'
```

To retrieve all: 
```
kubectl -n kube-system get pods -l k8s-app=kube-dns -o jsonpath='{range .items[]}{.metadata.name}{"\n"}{end}'
```

File: daemon/k8s_watcher.go
Comment (eloycoto) [inline daemon/k8s_watcher.go:1921]: That `undefined` should be a const no?

File: examples/kubernetes/addons/etcd-operator/tls/certs/gen-cert.sh
Comment (eloycoto) [inline examples/kubernetes/addons/etcd-operator/tls/certs/gen-cert.sh:6]: Should we add this into the contributing guide as a dependency?

**Review by eloycoto — APPROVED** [review]

File: pkg/node/manager.go
Comment (raybejjani) [inline pkg/node/manager.go]: Can source == FromAgentLocal be overwritten by any other source? We don't end up calling `UpdateNode` in `updateK8sNodeTunneling` but checking here makes the relationships between sources more clear.

File: daemon/k8s_watcher.go
Comment (raybejjani) [reply daemon/k8s_watcher.go:1921]: and also test for the string being empty?

**Review by raybejjani — COMMENTED** [review]
I couldn't get the readme instructions to work. I built cilium-agent locally, and modified the spec to use that file (mounted in at a specific path).
I could only get the cilium-etcd to start when I deleted cilium-development.yaml, but then cilium-development.yaml seemed to bounce between being ready and not. It also seemed like cilium status was blocked on something sometimes.

**Comment by raybejjani** [comment]
Following up on my comment. I tested with a prebuilt image for this PR (in case my hacky way was the problem). I did find that, on minikube, one needs to edit `cilium-etcd-cluster.yaml` to have a size of 1 otherwise etcd never comes and up and nothing works. I'm not sure if it's worth mentioning that somewhere since inevitably someone will try this on a <3 node system.

File: examples/kubernetes/addons/etcd-operator/cilium-development.yaml
Comment (ianvernon) [inline examples/kubernetes/addons/etcd-operator/cilium-development.yaml:315]: We can remove this yes?

File: examples/kubernetes/addons/etcd-operator/tls/certs/gen-cert.sh
Comment (ianvernon) [reply examples/kubernetes/addons/etcd-operator/tls/certs/gen-cert.sh:6]: I'd rather we put it into a docker image and have some type of wrapper script which generates stuff within that image so that we don't have to worry about maintaining more dependencies outside of images.

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (ianvernon) [inline examples/kubernetes/addons/etcd-operator/README.md]: "on" --> "by"

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (ianvernon) [inline examples/kubernetes/addons/etcd-operator/README.md]: "running." --> "running:"

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (ianvernon) [inline examples/kubernetes/addons/etcd-operator/README.md]: kube-dns --> \`\`kube-dns\`\`

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (ianvernon) [reply examples/kubernetes/addons/etcd-operator/README.md]: I'd make a note of this then in the "Deploy kube-dns..." step above. As a user who is following the guide, seeing restarts may be worrisome.

File: daemon/k8s_watcher.go
Comment (ianvernon) [inline daemon/k8s_watcher.go:1919]: Instead of FIXME create a GH issue and note it here in the format "GH-XXXX"

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: examples/kubernetes/addons/etcd-operator/cilium-development.yaml
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/cilium-development.yaml:315]: Let's drop this once we remove the support for 1.7

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/README.md]: Done

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/README.md]: Done

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/README.md]: Done

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/README.md:59]: It won't help. The etcd-operator will take some time until it creates the CRD. I'll create a file with the CRD and add it to the directory.

File: examples/kubernetes/addons/etcd-operator/README.md
Comment (aanm) [reply examples/kubernetes/addons/etcd-operator/README.md]: Done

File: daemon/k8s_watcher.go
Comment (aanm) [reply daemon/k8s_watcher.go:1919]: This is a copy from another part of the code.

File: daemon/k8s_watcher.go
Comment (aanm) [reply daemon/k8s_watcher.go:1921]: This is a copy from another part of the code.

File: daemon/k8s_watcher.go
Comment (aanm) [reply daemon/k8s_watcher.go]: Fixed

File: pkg/node/manager.go
Comment (aanm) [reply pkg/node/manager.go]: Done

Comment (aanm) [reply pkg/node/manager.go]: Ok

**Review by aanm — COMMENTED** [review]
@tgraf Storing the node's source in the node structure was tricky to the fact I have to force it to be `FromKVStore` in `OnUpdate()`

**Comment by aanm** [comment]
test-me-please

**Review by raybejjani — APPROVED** [review]

File: examples/kubernetes/addons/etcd-operator/tls/certs/gen-cert.sh
Comment (tgraf) [reply examples/kubernetes/addons/etcd-operator/tls/certs/gen-cert.sh:6]: This is a dependency on the user system.

* We probably don't want to require being able to run containers on the system that is running kubectl.
* There may be some resistance to running a random container image to generate certificates.

We can definitely offer it as an alternative but I would not make it the default.

**Review by tgraf — APPROVED** [review]

**Pull request merged**
